### PR TITLE
HybridListeners mutliple SslConfig support 

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

HybridListeners support defining multiple ssl configs for a sub-listener, instead of only supporting a single one

# Context

### How do we support this for HttpGateways?
A Gateway can be of type HttpGateway. In this instance, if the Gateway is toggled to support ssl, it only selects VirtualServices with SslConfigs and then aggregates those in a list on the Proxy.Listener. Then when we generate the FilterChain for this listener, we create duplicate FilterChains, each with a separate FilterChainMatch/TransportSocket based on the SslConfig. This makes it extremely easy for teams to maintain VirtualServices and just have the routing/filter chain logic support their ssl settings dynamically.

### What was the previous behavior?
Previously, a HybridGateway would define a Matcher(with SslConfig) and an HttpGateway (with VirutalServices). All VirtualServices selected by the HttpGateway were aggregated, but instead of aggregating the SslConfig from those VirtualServices, the Matcher.SslConfig was used for all routes. 
This was helpful in that the configuration was more explicit and transparent to the user. However, it removes the powerful capability of a team (maintaining a VS) being able to configure the SslConfig of their choosing. It meant that if teams wanted to support different Ssl settings, they would need to create different matcher+httpGateway combination. This would't be bad per se, just more verbose and different from the behavior they're used to.

### Why is this new behavior preferred?
This new behavior follows the pattern that we've established for HttpGateways.

If we want to support both behaviors in the future, we could expose a new API that allows users to flag whether to inherit the virtual service Ssl config, or override it. However, we don't need to introduce that feature yet, as it has not been requested.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

# Remaining Work
- [x] Kube2e tests to validate usage of feature
- [ ] Docs to walk through feature and new CRD

### CODE REVIEW

File: projects/gloo/pkg/translator/filter_chain.go [hunk 1]
```diff
@@ -55,63 +57,126 @@ func (t *tcpFilterChainTranslator) ComputeFilterChains(params plugins.Params) []
...
+		// nothing to do
+		return
+	}
+
+	if filterChain.GetFilterChainMatch() == nil {
+		// create a FilterChainMatch, if necessary
+		filterChain.FilterChainMatch = &envoy_config_listener_v3.FilterChainMatch{}
+	}
+
+	var envoySourcePrefixRanges []*envoy_config_core_v3.CidrRange
```
Comment (nfuden) [inline projects/gloo/pkg/translator/filter_chain.go:168]: nit: Might as well make this with size len(sourcePrefixRanges)

File: projects/gloo/pkg/translator/ssl_configuration.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	for _, v := range newValues {
+		if _, ok := existingValues[v]; !ok {
+			values = append(values, v)
+		}
+	}
+	return values
+}
+
+// Inspired by: https://github.com/solo-io/gloo/blob/0ad2a02a816be2b4a8b6ce27ff9db01206ce6ceb/projects/gateway/pkg/translator/merge_options.go#L10
+// I opted to copy/paste the code since it lives in a different project
```
Comment (nfuden) [inline projects/gloo/pkg/translator/ssl_configuration.go:55]: micro-nit: refrain from using I in the comments

File: projects/gloo/pkg/translator/ssl_configuration.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	for i := 0; i < dstValue.NumField(); i++ {
+		dstField, srcField := dstValue.Field(i), srcValue.Field(i)
+		shallowMerge(dstField, srcField, false)
+	}
+
+	return dst
+}
+
+// Sets dst to the value of src, if src is non-zero and dest is zero-valued or overwrite=true.
+func shallowMerge(dst, src reflect.Value, overwrite bool) {
```
Comment (nfuden) [inline projects/gloo/pkg/translator/ssl_configuration.go:79]: Thought: While I think from a time perspective this is a great solution I am a little worried that we are making our own shallow merges for these protobuf message. Unfortunately not doing this in a proto would likely be a large undertaking but given that I have found several discussions where people are saying that protoc gen does not create shallow merge functions is that it is unsafe to perform this action. 
We get away with it because we are creating a new base proto and copying it over but in general it sounds like we may be at odds with the designers of our protoc gen libraries.

**Review by nfuden — COMMENTED** [review]
Overall this seems reasonable and as far as I can trace does what it sets out to do.
That being said our copying and merging of different protos seems to only be growing as we add these higher order constructs. Given this I think in the future we should consider having shallow merge code rather than over leveraging reflection as reflection is slow. 

TLDR: This should be good once the tests are expanded on!

File: projects/gloo/pkg/translator/ssl_configuration.go (hunk 3 above)
Comment (nfuden) [reply projects/gloo/pkg/translator/ssl_configuration.go:79]: It is true that its safe in our case. Most of the discussions are around mutexs and other message sending context that we shouldnt be operating on

File: projects/gloo/pkg/translator/listener_subsystem.go [hunk 4]
```diff
@@ -156,51 +159,100 @@ func (l *ListenerSubsystemTranslatorFactory) GetHybridListenerTranslators(ctx co
 	hybridListenerReport := listenerReport.GetHybridListenerReport()
 	if hybridListenerReport == nil {
 		contextutils.LoggerFrom(ctx).DPanic("internal error: listener report was not hybrid type")
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (nfuden) [reply projects/gloo/pkg/translator/listener_subsystem.go:162]: Dpanic is only when in dev mode (which I guess we never unset). How would we feel about not panicking and bubbling up the error in the future?

File: projects/gloo/pkg/translator/ssl_configuration_test.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/gloo/projects/gloo/pkg/translator"
+)
+
+var _ = Describe("MergeSslConfig", func() {
+
+	It("merges top-level SslConfig fields", func() {
```
[the remark this answers is not in the document]
Comment (nfuden) [reply projects/gloo/pkg/translator/ssl_configuration_test.go:12]: If we are doing that we might as well just make a custom overwriting function right?
Or do you think a table test would fit well here

File: projects/gloo/pkg/translator/listener_subsystem.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bewebi) [reply projects/gloo/pkg/translator/listener_subsystem.go:162]: I missed that it was `DPanic`, I'm fine with it as is

File: projects/gloo/pkg/translator/ssl_configuration_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bewebi) [reply projects/gloo/pkg/translator/ssl_configuration_test.go:12]: Seems like it would lend well to a table test
I'm not sure I follow what you mean about a custom function...do you mean something that explicitly operates on `SslConfig` fields rather than the type-agnostic way `shallowMerge` does? Either way exhaustive testing for our use case is ideal

[the remark this answers is not in the document]
Comment (nfuden) [reply projects/gloo/pkg/translator/ssl_configuration_test.go:12]: ya exactly. If we are going through all the fields it seems like it might be faster/ better to iterate over the known fields rather than exercising reflection to figure out what fields exist. 
The approach via reflection is nicely generalizable (hence the json usage) but is bound to be massively slower

[the remark this answers is not in the document]
Comment (nfuden) [reply projects/gloo/pkg/translator/ssl_configuration_test.go:12]: If we dont hook the merge code into our auto-gen though we kinda need this agnostic approach to make sure that we dont miss new fields.

**Comment by nfuden** [comment]
LGTM

**Review by nfuden — APPROVED** [review]
Great work!
